From f002e71ba2282fa7096a9d7756a1f7e6e72cee19 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 09 Nov 2022 17:01:03 +0200
Subject: [PATCH] [SW-110085] arc: introduce wevt command

The wevt instruction is similar to sleep, but with several differences
that make it much more like the ARM's wfe intruction.
For example, it can be called from user mode (not only from kernel mode)
and it also allows the core CPU wake up on a wire event (like ARM's
wfe). In Greco/Gaudi2 the wake up event is supported through the
WAKE_UP_EVENT register in PSOC_ARCx_CFG.
Since no longer used, remove Gaudi2 sleep() macro.

Change-Id: I130d836b9ffe90a93a8bc7f7747fe32c2071cab9
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 9f220a9..4d8723f 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -269,10 +269,9 @@
 	if (!valid_arc_img) {
 		/*
 		 * Dont run ARC0 or jump to ARC1 ICCM if no arc images
-		 * received as part of fit. Replace sleep() with wfe()
-		 * once wfe() for gaudi2 is fixed.
+		 * received as part of fit.
 		 */
-		sleep();
+		wfe();
 	}
 
 	/* Set ARC0 in running state */
@@ -701,7 +700,7 @@
 	pcie_config();
 
 	/* Put secondary CPU in sleep state until further notice */
-	sleep();
+	wfe();
 }
 
 void secondary_main(void)
@@ -709,7 +708,7 @@
 	preboot_pcie_monitoring();
 
 	/* Put secondary CPU in sleep state until further notice */
-	sleep();
+	wfe();
 }
 
 void preboot_pcie_wait_for_link(void)
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index cbf43cb..64e7d00 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -127,8 +127,6 @@
 /* Only BP1 == 0x2 - means 2 blocks - up to 128KB (eROM) */
 #define PROT_MASK	(READ_STATUS_BP1 | READ_STATUS_SRWD)
 
-#define sleep()	__asm__ __volatile__("sleep\n\t" : : )
-
 void init_secondary(void);
 extern void secondary_entry(void);
 #endif /* __ASSEMBLY__ */
diff --git a/include/preboot.h b/include/preboot.h
index 9503f22..eac31ec 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -15,7 +15,7 @@
 #define ROUND_UP4(n)		(((n) + sizeof(u32) - 1) & ~(sizeof(u32) - 1))
 
 #ifdef __arc__
-#define wfe()	__asm__ __volatile__("sleep 0x11\n\t" : : )
+#define wfe()	__asm__ __volatile__("wevt\n\t" : : )
 #endif
 
 #ifdef __aarch64__
